messages/README.md

```diff
@@ -0,0 +1,108 @@
+
+As a result, simple `serde` tagged serialization/deserialization is not sufficient. We instead dedicate the `msg_types` module for this purpose.
+
+### The `msg_types` module
+We want to take the `@type` field and parse it to determine the exact protocol, version and message to process. The machinery is in place for that to happen through the delayed serialization/deserialization, but the protocol version resolution and all the possible variants for the `@type` field are located within
+this module and encapsulated in the `Protocol` enum.
+
+Similarly to `AriesMessage`, the `Protocol` enum is represented in a cascading fashion such as:
+
+- Protocol
+    - ConnectionType
+        - ConnectionTypeV1
+            - V1_0 -> ConnectionTypeV1_0
+                - Request
+                - Response
+            - V1_1 -> ConnectionTypeV1_1
+                - Request
+                - Response
+
+When deciding the minor version to use in a protocol, the protocol and version of the `@type` field is looked up in the `PROTOCOL_REGISTRY` (a lazily initialized map containing all protocols implemented). On success, the specified protocol version can be used as it is implemented. On failure, though, the minor version is decremented and looked up again until either the minor version reaches `0` or the lookup succeeds.
+
+**NOTE:** The `Protocol` enum has values like: `Protocol::ConnectionType::ConnectionTypeV1::V1_0`. The exact message kind is not part of the enum but rather there is type-linking involved (hence the arrow ->). This allows the `Protocol` enum to represent only protocols and protocols version by itself, while also providing mechanisms for parsing the message kind and thus getting the exact message to further process.
+
+### The `msg_fields` module
+This module contains the actual messages data structures with all fields defined (apart from `@type`).
+In practice, the module exports aliases for concrete definitions of `MsgParts`,
+such as ```pub type IssueCredential = MsgParts<IssueCredentialContent, IssueCredentialDecorators>;```. 
+
+The relevant submodule then contains the actual definitions of the `content` and `decorators` data structure (such as `IssueCredentialContent` and `IssueCredentialDecorators`).
+
+### The `decorators` module
+Contains data structures for decorators. 
+
+Unlike messages and their `@type` field, decorators get their version associated within their name `~thread/1`. Since only major versions are used, swapping a decorators version in a message represents a breaking change and would have to be explicitly defined in the message's content/decorators data structure. No resolution is required.
+
+### Extending the crate
+Adding new messages to the crate should be fairly easy to do, even without understanding all the inner workings. The main concepts needed are:  
+- the `AriesMessage` enum encapsulates all messages
```

Comment:
Another nitpick: 
the last character should be either `,` and continue the sentence in the next point, or keep each point a sentence starting with capital, finished with `.`

Reply:
Not really sure what you mean, but made me notice some of these points ended in periods. Fixed that.